File: usage_example/c/step2/src/main.c
#include <stdio.h>
#include "c3ga.h"

#define BUF_LEN 1024
/*
This small example creates a line and a plane from points.
It then computes the intersection point of the line and the plane.

In this step, all GA variables are stored in 'mv' (the multivector type).
*/
int main(int argc, char *argv[]) {

	normalizedPoint linePt1, linePt2, planePt1, planePt2, planePt3;
	line L, tmpL;
	plane P;
	flatPoint fp, intersection;
	dualLine dl;
	mv tmp1;
	char buf[BUF_LEN];

	// get five points
	cgaPoint_float_float_float(&linePt1, 1.0f, 0.0f, 0.0f);
	cgaPoint_float_float_float(&linePt2, 1.0f, 1.0f, 0.0f);
	cgaPoint_float_float_float(&planePt1, 1.0f, 2.0f, 0.0f);
	cgaPoint_float_float_float(&planePt2, 1.0f, 2.0f, 1.0f);
	cgaPoint_float_float_float(&planePt3, 0.0f, 2.0f, 1.0f);

	// output text the can be copy-pasted into GAViewer
	// warning: c_str() is not multi-threading safe (for that, use toString() instead).
	printf("linePt1 = %s,\n", toString_mv(normalizedPoint_to_mv(&tmp1, &linePt1), buf, BUF_LEN, "%2.2f")); 
	printf("linePt2 = %s,\n", toString_mv(normalizedPoint_to_mv(&tmp1, &linePt2), buf, BUF_LEN, "%2.2f"));
	printf("planePt1 = %s,\n", toString_mv(normalizedPoint_to_mv(&tmp1, &planePt1), buf, BUF_LEN, "%2.2f"));
	printf("planePt2 = %s,\n", toString_mv(normalizedPoint_to_mv(&tmp1, &planePt2), buf, BUF_LEN, "%2.2f"));
	printf("planePt3 = %s,\n", toString_mv(normalizedPoint_to_mv(&tmp1, &planePt3), buf, BUF_LEN, "%2.2f"));

	// create line and plane out of points
	op_normalizedPoint_flatPoint(&L, &linePt1, op_normalizedPoint_ni_t(&fp, &linePt2, &ni));
	op_normalizedPoint_line(&P, &planePt1, op_normalizedPoint_flatPoint(&tmpL, &planePt2, op_normalizedPoint_ni_t(&fp, &planePt3, &ni)));

	// output text the can be copy-pasted into GAViewer
	printf("L = %s,\n", toString_mv(line_to_mv(&tmp1, &L), buf, BUF_LEN, "%2.2f")); 
	printf("P = %s,\n", toString_mv(plane_to_mv(&tmp1, &P), buf, BUF_LEN, "%2.2f")); 

	// compute intersection of line and plane
	lc_dualLine_plane(&intersection, dual_line(&dl, &L), &P);

	printf("intersection = %s,\n", toString_mv(flatPoint_to_mv(&tmp1, &intersection), buf, BUF_LEN, "%2.2f"));

	return 0;
}

File: usage_example/java/Step2/src/Main.java
import C3ga_pkg.C3ga;
import C3ga_pkg.FlatPoint;
import C3ga_pkg.Line;
import C3ga_pkg.NormalizedPoint;
import C3ga_pkg.Plane;


public class Main {

    /**
    This small example creates a line and a plane from points.
    It then computes the intersection point of the line and the plane.

    In this step, all GA variables are stored in specialized multivector types,
    such as 'line', 'plane' and 'flatPoint'.
    */
	public static void main(String[] args) {
		
        // get five points
        NormalizedPoint linePt1 = C3ga.cgaPoint(1.0f, 0.0f, 0.0f);
        NormalizedPoint linePt2 = C3ga.cgaPoint(1.0f, 1.0f, 0.0f);
        NormalizedPoint planePt1 = C3ga.cgaPoint(1.0f, 2.0f, 0.0f);
        NormalizedPoint planePt2 = C3ga.cgaPoint(1.0f, 2.0f, 1.0f);
        NormalizedPoint planePt3 = C3ga.cgaPoint(0.0f, 2.0f, 1.0f);

        // output text the can be copy-pasted into GAViewer
        System.out.println("linePt1 = " + linePt1 + ",");
        System.out.println("linePt2 = " + linePt2 + ",");
        System.out.println("planePt1 = " + planePt1 + ",");
        System.out.println("planePt2 = " + planePt2 + ",");
        System.out.println("planePt3 = " + planePt3 + ",");

        // create line and plane out of points
        Line L = new Line(C3ga.op(linePt1, C3ga.op(linePt2, C3ga.ni)));

        Plane P = new Plane(C3ga.op(planePt1, C3ga.op(planePt2, C3ga.op(planePt3, C3ga.ni))));

        // output text the can be copy-pasted into GAViewer
        System.out.println("L = " + L + ",");
        System.out.println("P = " + P + ",");

        // compute intersection of line and plane
        FlatPoint intersection = new FlatPoint(C3ga.lc(C3ga.dual(L), P));

        System.out.println("intersection = " + intersection + ",");
	}

}
